tensorflow_datasets/core/lazy_imports_lib.py

```diff
@@ -20,24 +20,98 @@
 from __future__ import division
 from __future__ import print_function
 
+import contextlib
 import importlib
+import sys
+import types
 
 from tensorflow_datasets.core.utils import py_utils as utils
 
+_ERR_MSG = ("Failed importing {name}. This likely means that the dataset "
+            "requires additional dependencies that have to be "
+            "manually installed (usually with `pip install {name}`). See "
+            "setup.py extras_require.")
+
+# `lazy_imports` currently only allow the following external dependencies to
+# be used during dataset generation.
+_ALLOWED_LAZY_DEPS = [
+    "apache_beam",
+    "crepe",
+    "cv2",
+    "h5py",
+    "langdetect",
+    "librosa",
+    "matplotlib",
+    "matplotlib.pyplot",
+    "mwparserfromhell",
+    "nltk",
+    "pandas",
+    "PIL_Image",
+    "PIL.TiffImagePlugin",
+    "pretty_midi",
+    "pydub",
+    "scipy",
+    "scipy.io",
+    "scipy.ndimage",
+    "skimage",
+    "skimage.color",
+    "skimage.filters",
+    "skimage.external.tifffile",
+    "tensorflow_io",
+    "tldextract",
+    "os",
+    # "test_foo",
+]
+
 
 def _try_import(module_name):
   """Try importing a module, with an informative error message on failure."""
   try:
     mod = importlib.import_module(module_name)
     return mod
   except ImportError:
-    err_msg = ("Failed importing {name}. This likely means that the dataset "
-               "requires additional dependencies that have to be "
-               "manually installed (usually with `pip install {name}`). See "
-               "setup.py extras_require.").format(name=module_name)
+    err_msg = _ERR_MSG.format(name=module_name)
     utils.reraise(suffix=err_msg)
 
 
+class FakeModule(types.ModuleType):
+  """A fake module which raise ImportError whenever an unknown attribute is accessed"""
+  def __init__(self, name):
+    self.__path__ = None
+    super(FakeModule, self).__init__(name)
+
+  def __getattr__(self, attr):
+    err_msg = _ERR_MSG.format(name=self.__name__)
+    raise ImportError(err_msg)
+
+
+class LazyImporterHook(object):
```

Comment:
If I understand the implementation correctly, if executing:

```py
with lazy_imports():
  import x
```
And that `x` internally imports `y`:
```py
# x/__init__.py
import y
```
I'm a little concerned that `y` gets lazy imported even if this is internal code that TFDS shouldn't modify.
Do you think it would be possible to make the importer such as only the top-most level gets lazy imported ?

Reply:
I will look into this.

Reply:
After thinking about this, I think it may be better to patch the built-in `__import__()` directly.
* The importer hook implementation seems too complicated to correctly handle edge cases.
* We do not want to add the lazy imports to `sys.modules`, so it seems we're breaking some assumptions of the hook protocol

You can use `with py_utils.temporary_assignment(built_in, '__import__', new_import):` to patch temporary the built-in.

Reply:
I can definitely do this.
However, if please review the new implementation using `sys.meta_path`.
If you think this is complicated, I will switch to the suggested method

Reply:
The issue with `sys.meta_path` is that it won't handle correctly the nested case (lazy-imports inside the import). Also, the lazy imports shouldn't be registered in `sys.modules` as it might break others libs.

We should add a test to make sure that nested work correctly:

```py
with self.assertRaisesRegex(ImportError, 'y not found'):
  with maybe_imports():
    import x
```
With:
* ALLOWED_DEPS is x, y
* x is installed, but not y
* `x.py`:
```py
# x.py
import y
# y should raise ImportError, instead of being a dummy module
```